bidict/_bidict.py

```diff
@@ -110,7 +110,7 @@ def pop(self, __key: KT, __default: DT) -> t.Union[VT, DT]: ...
     @t.overload
     def pop(self, __key: KT) -> VT: ...
     @t.overload
-    def pop(self, __key: KT, __default: t.Union[VT, DT] = ...) -> t.Union[VT, DT]: ...
+    def pop(self, __key: KT, __default: t.Any = ...) -> t.Any: ...
```

Comment:
Your overload is a bit squirrelly here. I think pyright is correct in pointing out the incompatible overlap. I'm not sure why mypy doesn't flag it in this case.

If I correctly understand the intent of this overload, I think this would be a better solution:

```python
    @t.overload
    def pop(self, __key: KT, __default: t.Literal[MISSING]) -> VT: ...
    @t.overload
    def pop(self, __key: KT, __default: DT) -> t.Union[VT, DT]: ...
    @t.overload
    def pop(self, __key: KT) -> VT: ...
```

Reply:
Thanks, @erictraut, your proposed type hints indeed appease pyright, but now mypy is rejecting them:
```console
bidict/_bidict.py: note: In member "pop" of class "MutableBidict":
bidict/_bidict.py:108:6: error: Signature of "pop" incompatible with supertype "MutableMapping"  [override]
        @t.overload
         ^
bidict/_bidict.py:108:6: note:      Superclass:
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT) -> VT
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def [_T] pop(self, KT, Union[VT, _T] = ...) -> Union[VT, _T]
bidict/_bidict.py:108:6: note:      Subclass:
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT, Any) -> VT
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def [DT] pop(self, KT, DT) -> Union[VT, DT]
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT) -> VT
bidict/_bidict.py: note: In function "pop":
bidict/_bidict.py:109: error: Parameter 1 of Literal[...] is invalid  [misc]
        def pop(self, __key: KT, __default: 't.Literal[MISSING]') -> VT: ...
        ^
bidict/_bidict.py:109:41: error: Variable "bidict._typing.MISSING" is not valid as a type  [valid-type]
        def pop(self, __key: KT, __default: 't.Literal[MISSING]') -> VT: ...
                                            ^
bidict/_bidict.py:109:41: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

Also, `Literal` requires Python 3.8+, and `TypeAlias` requires 3.10+. Bidict currently supports 3.7+, and it'd be nice to avoid taking a dependency on typing_extensions.

Any other suggestions?

Reply:
You can change the first overload to accept `__default: MissingT`.

Reply:
I meant to say, I tried that too, and again, it made pyright happy but made mypy start failing, as follows:
```python
    @t.overload
    def pop(self, __key: KT, __default: MissingT) -> t.Any: ...
    @t.overload
    def pop(self, __key: KT, __default: DT) -> t.Union[VT, DT]: ...
    @t.overload
    def pop(self, __key: KT) -> VT: ...

    def pop(self, key: KT, default: ODT[DT] = MISSING) -> t.Union[VT, DT]:
        ...
```
```console
bidict/_bidict.py: note: In member "pop" of class "MutableBidict":
bidict/_bidict.py:108:6: error: Signature of "pop" incompatible with supertype "MutableMapping"  [override]
        @t.overload
         ^
bidict/_bidict.py:108:6: note:      Superclass:
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT) -> VT
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def [_T] pop(self, KT, Union[VT, _T] = ...) -> Union[VT, _T]
bidict/_bidict.py:108:6: note:      Subclass:
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT, MissingT) -> Any
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def [DT] pop(self, KT, DT) -> Union[VT, DT]
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT) -> VT
```
Is this a known issue between pyright and mypy?

Reply:
I'll leave it to you to debug the mypy issues here. The original overload was overlapping, and pyright was correct to flag it as such. You can decide how you'd like to fix it. If you suspect there are bugs in mypy, please report those to the mypy maintainers.

Reply:
@erictraut For sure, my question was as much for @JelleZijlstra (and anyone else who might see it and know how to help) as it was for you. I really appreciate all the help you've been giving here, and if you ever feel you don't have time to continue, I understand completely. I hope the bugs in pyright that this process has uncovered has made it at least somewhat worth it, and I appreciate any additional thought you can give to my usability suggestions.

Reply:
With the type hints for this method in the latest revision of this PR, pyright no longer flags any problems, but now mypy does:
```console
bidict/_bidict.py: note: In member "pop" of class "MutableBidict":
bidict/_bidict.py:108:6: error: Signature of "pop" incompatible with supertype "MutableMapping"  [override]
        @t.overload
         ^
bidict/_bidict.py:108:6: note:      Superclass:
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT) -> VT
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def [_T] pop(self, KT, Union[VT, _T] = ...) -> Union[VT, _T]
bidict/_bidict.py:108:6: note:      Subclass:
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT) -> VT
bidict/_bidict.py:108:6: note:          @overload
bidict/_bidict.py:108:6: note:          def pop(self, KT, VT) -> VT
Found 1 error in 1 file (checked 14 source files)
```

I'm not sure if mypy, pyright, the `MutableMapping.pop` type hints, or I am the one who's wrong here. But since the latest type hints look best to me, I'm going to use a `# type: ignore [override]` here for now, and come back to this after @JelleZijlstra has a chance to take another look.

Reply:
I think mypy is right there. The superclass allows you to pass an unrelated type as the second argument, but the subclass doesn't.

Pyright doesn't seem to handle overrides involving overloads correctly. I'll report a bug.

Reply:
On line 111 of the latest revision, the subclass’s overload does allow you to pass an unrelated type:
```python
def pop(self, __key: KT, __default: DT) -> t.Union[VT, DT]: ...
```
No?

Reply:
Do you still get the same mypy error from that? I was going off the error in your previous message.

Reply:
Yes I do. Sorry that wasn’t clear.

Reply:
Filed https://github.com/python/mypy/issues/12390 so this bug can be tracked properly.